### PR TITLE
lightning: fix panic due to row's column exceeds definition

### LINKED ISSUE #33396: lightning panic due to column count mismatch
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

create table with x columns
import csv with lightning with columns > x
### 2. What did you expect to see? (Required)
lightning report error
### 3. What did you see instead (Required)
```
panic: runtime error: index out of range [265] with length 265
goroutine 275 [running]:
github.com/pingcap/tidb/br/pkg/lightning/backend/tidb.(*tidbEncoder).Encode(0xc0000b3d60, 0xc00168c060, 0xc001d0e000, 0xe4ad, 0xe4ad, 0x1, 0xc0002b4000, 0x109, 0x10a, 0xc000066abe, ...)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/backend/tidb/tidb.go:321 +0xfc5
github.com/pingcap/tidb/br/pkg/lightning/restore.(*chunkRestore).encodeLoop(0xc0005a6940, 0x4760048, 0xc0002d1740, 0xc00168c000, 0xc0004db800, 0xc00168c060, 0x4736a58, 0xc0000b3d60, 0xc00005d200, 0xc000991900, ...)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/restore/restore.go:2389 +0x3ad
github.com/pingcap/tidb/br/pkg/lightning/restore.(*chunkRestore).restore(0xc0005a6940, 0x4760048, 0xc0002d1740, 0xc0004db800, 0x0, 0xc0005a6960, 0xc0005a6980, 0xc000991900, 0x0, 0x0)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/restore/restore.go:2484 +0x713
github.com/pingcap/tidb/br/pkg/lightning/restore.(*TableRestore).restoreEngine.func3(0xc000524f40, 0xc000991900, 0x3ff0000000000000, 0x4760048, 0xc0002d1740, 0xc0004db800, 0xc000000000, 0xc0005a6960, 0xc0005a6980, 0xc000524f28, ...)
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/restore/table_restore.go:542 +0x17e
created by github.com/pingcap/tidb/br/pkg/lightning/restore.(*TableRestore).restoreEngine
        /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/restore/table_restore.go:534 +0x9a9
```

### PR DESCRIPTION
**Pull request by buchuitoudegou**
### What problem does this PR solve?

Issue Number: close #33396

Problem Summary:

### What is changed and how it works?
Check if the number of columns exceeds that of the permutation, which should equal to the table definition, before populating the insert statement.

### Check List

Tests 

- [x] Unit test

Side effects

Documentation

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 1]
```diff
@@ -594,6 +595,14 @@ func encodeRowsTiDB(t *testing.T, b backend.Backend, tbl table.Table) kv.Rows {
...
 
 	row.ClassifyAndAppend(&dataRows, &dataChecksum, &indexRows, &indexChecksum)
+
+	rawRow := make([]types.Datum, 0)
+	for i := 0; i < 15; i++ {
+		rawRow = append(rawRow, types.NewIntDatum(0))
+	}
+	row, err = encoder.Encode(logger, rawRow, 1, []int{0, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1}, "12.csv", 0)
+	require.NotNil(t, err)
+	require.Regexp(t, regexp.MustCompile(".*column count mismatch, at most.*"), err.Error())
```
Comment (gozssky) [inline br/pkg/lightning/backend/tidb/tidb_test.go:605]:
```suggestion
	require.Regexp(t, "column count mismatch, at most", err.Error())
```

Comment (gozssky) [reply br/pkg/lightning/backend/tidb/tidb_test.go:605]: Or just `require.Contains` is enough?

**Review by gozssky — COMMENTED** [review]
rest LGTM

File: br/pkg/lightning/backend/tidb/tidb_test.go (hunk 1 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/backend/tidb/tidb_test.go]: Have fixed it.

**Review by gozssky — APPROVED** [review]

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 2]
```diff
@@ -594,6 +594,14 @@ func encodeRowsTiDB(t *testing.T, b backend.Backend, tbl table.Table) kv.Rows {
 	require.NoError(t, err)
 
 	row.ClassifyAndAppend(&dataRows, &dataChecksum, &indexRows, &indexChecksum)
+
+	rawRow := make([]types.Datum, 0)
```
Comment (ngaut) [inline br/pkg/lightning/backend/tidb/tidb_test.go:598]: Make slice cap with 15.

Comment (ngaut) [inline br/pkg/lightning/backend/tidb/tidb_test.go:598]: Make slice cap with 15.

Comment (buchuitoudegou) [reply br/pkg/lightning/backend/tidb/tidb_test.go:598]: Will fix it.

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 3]
```diff
@@ -594,6 +594,14 @@ func encodeRowsTiDB(t *testing.T, b backend.Backend, tbl table.Table) kv.Rows {
 	require.NoError(t, err)
 
 	row.ClassifyAndAppend(&dataRows, &dataChecksum, &indexRows, &indexChecksum)
+
+	rawRow := make([]types.Datum, 14)
+	for i := 0; i < 15; i++ {
+		rawRow = append(rawRow, types.NewIntDatum(0))
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/tidb/tidb_test.go:600]: rawRow will have 29 items, but it's ok for test.

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 4]
```diff
@@ -594,6 +594,14 @@ func encodeRowsTiDB(t *testing.T, b backend.Backend, tbl table.Table) kv.Rows {
 	require.NoError(t, err)
 
 	row.ClassifyAndAppend(&dataRows, &dataChecksum, &indexRows, &indexChecksum)
+
+	rawRow := make([]types.Datum, 14)
```
Comment (gozssky) [inline br/pkg/lightning/backend/tidb/tidb_test.go:598]: `rawRow := make([]types.Datum, 15)` ? 🤔

**Comment by Ehco1996** [comment]
/hold

wait br it test passed

**Comment by Ehco1996** [comment]
/unhlod  breaked CI passed
https://ci.pingcap.net/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/11185/pipeline/114/

**Comment by Ehco1996** [comment]
/run-unit-test

[FAIL]  planner/core TestHashPartitionPruning

**Pull request merged**